File: 2232-Minimize-Result-by-Adding-Parentheses-to-Expression.py
class Solution:
    def minimizeResult(self, expression: str) -> str:

        exp_list = expression.split('+')
        exp_part1 = exp_list[0]
        exp_part2 = exp_list[1]

        min_val = sys.maxsize
        min_key = ''

        for i in range(len(exp_part1)):
            for j in range(1, len(exp_part2) + 1):
                key = exp_part1[:i] + '(' + exp_part1[i:] + \
                '+' + exp_part2[:j] + ')' + exp_part2[j:]
                val = self.evaluate(key)

                if val < min_val:
                    min_val = val
                    min_key = key
        return min_key

    
    def evaluate(self, expression):
        
        index1 = expression.index('(')
        index2 = expression.index('+')
        index3 = expression.index(')')

        a2 = int(expression[index1 + 1:index2]) + int(expression[index2 + 1:index3])

        a1 = 1 if expression[:index1] == '' else int(expression[:index1])
        a3 = 1 if expression[index3 + 1:] == '' else int(expression[index3 + 1:])
        
        return a1 * a2 * a3

File: 2232-Minimize-Result-by-Adding-Parentheses-to-Expression.cpp
class Solution {
public:
    string minimizeResult(string expression) {
        int index = expression.find("+");
        string exp_part1 = expression.substr(0, index);
        string exp_part2 = expression.substr(index + 1, expression.length() - index);

        int min_val = INT_MAX;
        string min_res = "";

        for (int i = 0; i < exp_part1.length(); i++) {
            for (int j = 1; j < exp_part2.length() + 1; j++) {
                string key =         exp_part1.substr(0, i) + "(" + exp_part1.substr(i, exp_part1.length() - i) \
                             + "+" + exp_part2.substr(0, j) + ")" + exp_part2.substr(j, exp_part2.length() - j);
                int val = evaluate(key);

                if (val < min_val) {
                    min_val = val;
                    min_res = key;
                }
            }
        }

        return min_res;
    }

    int evaluate(string expression) {
        int index1 = expression.find("(");
        int index2 = expression.find("+");
        int index3 = expression.find(")");

        string a1_str = expression.substr(0, index1);
        string a2_str = expression.substr(index1 + 1, index2 - index1 - 1);
        string a3_str = expression.substr(index2 + 1, index3 - index2 - 1);
        string a4_str = expression.substr(index3 + 1, expression.length() - index3 - 1);

        int a1;
        int a2 = stoi(a2_str);
        int a3 = stoi(a3_str);
        int a4;

        if (a1_str.length() == 0) {
            a1 = 1;
        } else {
            a1 = stoi(a1_str);
        }
    
        if (a4_str.length() == 0) {
            a4 = 1;
        } else {
            a4 = stoi(a4_str);
        }

        return a1 * (a2 + a3) * a4;
    }
};
